File: 748. Shortest Completing Word/748.go
package  leetcode

import(
	"strings"
)

func shortestCompletingWord(licensePlate string, words []string) string {
	licenseDict := make(map[rune]int, 0)
	for _,v := range licensePlate{
		if 0<=v-'a' && v-'a'<=25{
			licenseDict[v-'a']++
		}
		if 0<=v-'A' && v-'A'<=25{
			licenseDict[v-'A']++
		}
	}

	res := "aaaaaaaaaaaaaaaa"
	for _,v := range words{
		tmp := make(map[rune]int, 0)
		for _,v1 := range v{
			if 0<=v1-'a' && v1-'a'<=25{
				tmp[v1-'a']++
			}
		}

		flag := 0
		for i, v2 := range licenseDict{
			if v2>tmp[i]{
				flag = 1
				break
			}
		}
		if flag == 0{
			if len(v)<len(res){
				res = v
			}
		}
	}

	if len(res) < 16{
		return res
	}
	return ""
}

func shortestCompletingWord1(licensePlate string, words []string) string {

    s := ""
	minLen := 16
	index := make([]int, 26, 26)
	tmp := strings.ToLower(licensePlate)
    
	for _, v := range tmp {
		if v >= 'a' && v <= 'z' {
			index[v-'a']++
		}
	}
	for _, v := range words {
		if len(v) >= minLen {
			continue
		}
		if !matches(index, v) {
			continue
		}
		minLen = len(v)
		s = v
	}

	return s

}

func matches(index []int, s string) bool {
	temp := make([]int, 26, 26)
	for _, v := range s {
		temp[v-'a']++
	}
	for i := 0; i < 26; i++ {
		if index[i] > temp[i] {
			return false
		}
	}
	return true
}

File: 748. Shortest Completing Word/_748.py
class Solution:
    def shortestCompletingWord(self, licensePlate: str, words) -> str:
        licenseDict = {}
        for i in licensePlate:
            if 96<ord(i)<123:
                v = licenseDict.get(ord(i), 0)
                licenseDict[ord(i)] = v+1
            if 64<ord(i)<91:
                v = licenseDict.get(ord(i)+32, 0)
                licenseDict[ord(i)+32] = v+1

        res = list()
        for i in words:
            flag = 0
            for j in licenseDict:
                if licenseDict[j]>i.count(chr(j))+i.count(chr(j-32)):
                    flag = 1
                    break
            if flag == 0:
                res.append(i)
            
        res.sort(key = lambda i:len(i))
        return res[0]

    # 这代码真的是很Python
    # .lower()， .isalpha()来筛选元素
    # filter函数用的好
    # 下面一个列表生成式 里面的 & 用的很强啊 Counter特有啊， 学习了， 一直以为Counter和dict是一样的。。。
    # 然后min也用的很好，原来min也向sort一样可以加key函数
    def shortestCompletingWord1(self, licensePlate: str, words) -> str:
        from collections import Counter
        pc = Counter(filter(lambda x : x.isalpha(), licensePlate.lower()))
        return min([w for w in words if Counter(w) & pc == pc], key=len) 

    # 这个减法用的也是很有创意想法
    def shortestCompletingWord2(self, lp, words):
        import collections
        cnt = collections.Counter("".join(c for c in lp.lower() if c.isalpha()))
        return min([w for w in words if not cnt - collections.Counter(w)], key = len)

a = Solution()
print(a.shortestCompletingWord("1s3 456", ["looks", "pest", "stew", "show"]))